cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineRunner.java

```diff
@@ -76,10 +80,16 @@ public SparkStreamingPipelineRunner(JavaSparkExecutionContext sec, JavaStreaming
       source = sec.getPluginContext().newPluginInstance(stageInfo.getName(), macroEvaluator);
     }
 
+    DataTracer dataTracer = sec.getDataTracer(stageInfo.getName());
     StreamingContext sourceContext = new DefaultStreamingContext(stageInfo.getName(), sec, streamingContext);
-    JavaDStream<Object> javaDStream = source.getStream(sourceContext)
-      .transform(new CountingTranformFunction<>(stageInfo.getName(), sec.getMetrics(), "records.out",
-                                                sec.getDataTracer(stageInfo.getName())));
+    JavaDStream<Object> javaDStream = source.getStream(sourceContext);
+    CountingTranformFunction<Object> function = new CountingTranformFunction<>(stageInfo.getName(), sec.getMetrics(),
+                                                                               "records.out", dataTracer);
+    if (dataTracer.isEnabled()) {
+      javaDStream = javaDStream.transform(new LimitingFunction<>(function, numOfRecordsPreview));
+    } else {
```

Comment:
it would be better to just make the limiting function limit things instead of passing it another function. In other words:

```
  if (dataTracer.isEnabled()) {
    javaDStream = javaDStream.transform(new LimitingFunction(numOfRecordsPreview));
  }
  javaDStream = javaDStream.transform(function);
```

please also add a comment explaining why we need to call transform instead of filter.